Add tests for activity definition - Create, Delete, Edit

yash-learner opened the pull request:
## Proposed Changes

Fixes #14174 

Tagging: @ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate the bug or test the new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is placed in I18n files.
- [ ] Prepare a screenshot or demo video for the changelog entry and attach it to the issue.
- [ ] Request peer reviews.
- [ ] Complete QA on mobile devices.
- [ ] Complete QA on desktop devices.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Tests**
  * Added comprehensive test coverage for activity definition workflows (create, edit, delete operations).
  * Enhanced test infrastructure with new UI helpers for multi-select, value-set, and category pickers.
  * Optimized Playwright configuration for improved test reliability and debugging.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yash-learner committed:
wip: Add test for activity definition form

- Add two cases
- Add api-context util
- Add helper to get facility id and category id

yash-learner committed:
add alias

yash-learner committed:
refactor: update facility and resource category retrieval in tests

- Change `getFacilityAndCategory` to return a `FacilityRead` object instead of `facilityId`.
- Update helper functions to fetch the last facility and resource category because API returns in desc order
- Adjust test cases to use the new structure for accessing facility and resource category data.

yash-learner committed:
refactor: introduce selectFromCommand helper for streamlined UI interactions in tests

- Add `selectFromCommand` function to encapsulate logic for selecting options from command inputs.
- Refactor activity definition form test to utilize the new helper for improved readability and maintainability.

yash-learner committed:
Enhance activity definition form tests with comprehensive field validation

- Update `activity-definition-form.spec.ts` to include a new test for filling out the form with all fields and submitting.
- Refactor `selectFromCommand` and add `selectFromLocationMultiSelect` helper functions to improve UI interaction handling.
- Implement defensive checks to close any open popovers before interactions to ensure stability during tests.

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
fix: update API base URL in Playwright configuration

- Change `apiBaseURL` from `PLAYWRIGHT_API_BASE_URL` to `REACT_CARE_API_URL`

yash-learner committed:
Merge branch 'add-tests-for-labs' of https://github.com/ohcnetwork/care_fe into add-tests-for-labs

yash-learner committed:
wip: Organise specs

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Refactor: reorganize test helpers and setup for facility data

- Remove deprecated `helpers.ts` and introduce `helpers/helpers.ts` for improved modularity.
- Implement `loadFacility` and `getResourceCategory` functions to streamline facility and resource category retrieval.
- Add `facility.setup.ts` to prepare and cache facility data for tests, enhancing test reliability.
- Update `activity-definition-form.spec.ts` to utilize new helper functions for better test structure.

yash-learner committed:
Rename file and add cases for existing category

bodhish commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
Let use `tests/facility/settings/ instead of `tests/facility-settings`

& also camelCase? `activityDefinitions` Instead of `activity-definitions`

yash-learner replied:
I went with `kebab-case` because already present folders under playwright has `kebab-case`

for example `test-results`

even some llm mentioned playwright tests prefer `kebab-case`

---

Will update

yash-learner committed:
Remove explicit timeouts

yash-learner committed:
Fix typo in file name

yash-learner commented on tests/helpers/ui.ts:
```diff
@@ -0,0 +1,194 @@
+  closeAfterSelect?: boolean;
+}
+
+export async function selectFromCommand(
```
Later: I think this method can be improved, I remember adding some bad code because of my local backend causing timeouts and went into agent wrong suggestions, I think I removed most after realization but later when there is time test and fix any

yash-learner commented on tests/setup/facility.setup.ts:
```diff
@@ -0,0 +1,54 @@
+  // Save facility data to file
+  fs.writeFileSync(facilityFile, JSON.stringify(facility, null, 2));
+
+  console.log("Facility setup complete:", {
```
This console log can be removed

yash-learner commented on tests/facility-settings/activity-definitions/category-list.spec.ts:
```diff
@@ -0,0 +1,182 @@
+  setup = { facility };
+});
+
+test.describe("Activity Definition List", () => {
```
Fix this

bodhish commented on tests/setup/facility.setup.ts:
```diff
@@ -0,0 +1,54 @@
+import { createAuthenticatedAPIContext } from "@/tests/utils/auth-context";
```
A more simple approach without the overhead of maintaining another API layer. 

```ts
import { test, expect } from "@playwright/test";
import fs from "fs";

test.use({ storageState: '"tests/.auth/user.json" });

test("enter facility via UI and save facility id", async ({ page }) => {
  await page.goto("/");
 // the backend has a fixture for a fixed facility with everything configured for an happy flow
  await page.getByRole("link", { name: "Facility with Patient" }).click();
  await page.waitForURL(/\/facility\/([^/]+)\/home$/);
  const id = page.url().match(/\/facility\/([^/]+)\/home$/)?.[1];
  fs.mkdirSync("tests/.auth", { recursive: true });
  fs.writeFileSync("tests/.auth/facilityMeta.json", JSON.stringify({ id }, null, 2));
});
```

and a helper 

```ts
# tests/support/facilityId.ts
import fs from "fs";
import path from "path";
import { execSync } from "child_process";

const META_PATH = path.resolve("tests/.auth/facilityMeta.json");
let cachedId: string | null = null;

/**
 * Returns the facilityId saved during setup.
 * Auto-runs the setup if the meta file is missing or invalid.
 */
export function getFacilityId(): string {
  if (cachedId) return cachedId;

  if (!fs.existsSync(META_PATH)) {
    console.warn("⚠️ Facility meta missing — running setup.facility.ui.ts ...");
    execSync("npx playwright test tests/setup/facility.setip.ts", { stdio: "inherit" });
  }

  const raw = fs.readFileSync(META_PATH, "utf8");
  try {
    const { id } = JSON.parse(raw);
    if (!id) throw new Error("Missing id in facilityMeta.json");
    cachedId = id;
    return id;
  } catch (err) {
    throw new Error(`Invalid facility-meta.json: ${err instanceof Error ? err.message : err}`);
  }
}
```

and usage
```
import { getFacilityId } from "@test/support/facilityId";

const facilityId = getFacilityId();
```

yash-learner replied:
Yeah, thought of this approach as well while writing but went with API call 😁

because I want to each test to visit urls directly and want each tests to be focused and do not want to traverse from top level pages which makes more backend requests and more renders and more loading which can make tests flaky and increase test run time.

so I need categoryid and will be requiring other ID's in many other places so went with API call

yash-learner replied:
updated https://github.com/ohcnetwork/care_fe/pull/14108/commits/26b564151722a4ca62806e9ea477518384d4b660#diff-87b4416326c7a5903937c2a2ea409a7fde3e41a959150d3bf1c632106819c7ec, https://github.com/ohcnetwork/care_fe/pull/14108/commits/da3b13f88aae954177296c9571739d84948fd65e

bodhish commented on tests/utils/auth-context.ts:
```diff
@@ -0,0 +1,43 @@
+import playwrightConfig from "@/playwright.config";
```
I don't think we will need this

bodhish commented on tests/helpers/helpers.ts:
```diff
@@ -0,0 +1,65 @@
+import { createAuthenticatedAPIContext } from "@/tests/utils/auth-context";
```
Also lets remove this file

bodhish commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,205 @@
+export interface TestSetup {
+  facility: FacilityRead;
+  resourceCategory: ResourceCategoryRead;
+}
```
```suggestion
const facilityId = getFacilityId();
```

bodhish commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,205 @@
+    "activity_definition",
+  );
+  setup = { facility, resourceCategory };
+});
```
```suggestion
```

bodhish commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,205 @@
+test.describe("Activity Definition Form", () => {
+  test("should load the form and display page content", async ({ page }) => {
+    await page.goto(
+      `/facility/${setup.facility.id}/settings/activity_definitions/categories/${setup.resourceCategory.slug}/new`,
```
```suggestion
      `/facility/${facilityId}/settings/activity_definitions/`,
```

bodhish replied:
You can a pick a category after this that part of the fixtures.

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/14108#discussion_r2462287540

bodhish commented on tests/facility-settings/activity-definitions/activity-definition-form.spec.ts:
```diff
@@ -0,0 +1,205 @@
+      `/facility/${setup.facility.id}/settings/activity_definitions/categories/${setup.resourceCategory.slug}/new`,
+    );
+
+    await page.waitForLoadState("networkidle", { timeout: 30000 });
```
I don't think we need this

yash-learner replied:
Yes, we do not need 

I remove, cursor keeps adding

bodhish commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,205 @@
+
+    await expect(page.getByLabel(/title.*\*/i)).toBeVisible();
+    await expect(page.getByLabel(/slug.*\*/i)).toBeVisible();
+    await expect(page.getByLabel(/description.*\*/i)).toBeVisible();
```
why don't we fill with the label so that in 1 line both presence and fill can be tested.

yash-learner replied:
Wil fix it

bodhish commented on tests/facility-settings/activity-definitions/activity-definition-form.spec.ts:
```diff
@@ -0,0 +1,205 @@
+});
+
+test.describe("Activity Definition Form", () => {
+  test("should load the form and display page content", async ({ page }) => {
```
I don't think this test case is of any use as it would be tested in other sub cases.

yash-learner replied:
Got it, yes

yash-learner replied:
Removed https://github.com/ohcnetwork/care_fe/pull/14108/commits/26b564151722a4ca62806e9ea477518384d4b660#diff-87b4416326c7a5903937c2a2ea409a7fde3e41a959150d3bf1c632106819c7ec

yash-learner committed:
Improve naming a bit

bodhish commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,205 @@
+
+    await expect(page).toHaveURL(
+      `/facility/${setup.facility.id}/settings/activity_definitions`,
+    );
```
I would actually navigate to the activity definition and validate what I have added is what got saved.

yash-learner replied:
Yes, that needs to be checked, will update

yash-learner replied:
updated https://github.com/ohcnetwork/care_fe/pull/14108/commits/26b564151722a4ca62806e9ea477518384d4b660#diff-87b4416326c7a5903937c2a2ea409a7fde3e41a959150d3bf1c632106819c7ec

bodhish commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,205 @@
+    ).toBeVisible();
+
+    await expect(page).toHaveURL(
+      `/facility/${setup.facility.id}/settings/activity_definitions`,
```
I would actually navigate to the activity definition and validate what I have added is what got saved.

yash-learner replied:
Yes, that needs to be checked, will update

yash-learner replied:
Updated https://github.com/ohcnetwork/care_fe/pull/14108/commits/26b564151722a4ca62806e9ea477518384d4b660#diff-87b4416326c7a5903937c2a2ea409a7fde3e41a959150d3bf1c632106819c7ec

bodhish commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,205 @@
+      `/facility/${setup.facility.id}/settings/activity_definitions`,
+    );
+  });
+});
```
We need another test case that validates edit flow

yash-learner replied:
Yeah. work in progress

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
WIP: Make changes as per suggestions from Bodhi

- Update categoryList
- Remove API files
- use facility ID

yash-learner committed:
Update Playwright configuration and refactor activity definition tests

- Increased retry count for local in playwright.config.ts
- Refactored activity definition tests to use facility ID and improved selector functions
- Removed unused imports and streamlined test setup
- Introduced new helper functions for selecting from requirements and valuesets and category picker

yash-learner commented on playwright.config.ts:
```diff
@@ -17,22 +17,20 @@ export default defineConfig({
-  /* Retry on CI only */
-  retries: process.env.CI ? 2 : 0,
+
+  retries: process.env.CI ? 2 : 3,
```
what if  my system is short on cpu and memory with too many services and tools  running locally ?

yash-learner replied:
Also if it is flaky playwright report shows it

yash-learner committed:
Address some copilot comments

yash-learner committed:
Add test for creating nested category in activity definitions

yash-learner committed:
Add edit case and other refactors

yash-learner committed:
Remove the explicit timeout, to fix CI - Let PW handle it

https://github.com/ohcnetwork/care_fe/actions/runs/18927252776/job/54037864162?pr=14108

yash-learner committed:
Add delete AD test case and other changes

yash-learner committed:
Revert change made to user schedule creation

yash-learner committed:
Refactor and be specific with selectors

yash-learner commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,536 @@
+        }),
+      });
+      await expect(overviewCard).toBeVisible();
+      await expect(overviewCard.getByText(categoryName)).toBeVisible();
```
There are two fields

category and resourceCategory

yash-learner committed:
Add Playwright tests for activity definitions list functionality

- Fix back button for view page

yash-learner committed:
Fix some flakyness

yash-learner committed:
CI tests can run in parallel shards so create test can run before list

yash-learner committed:
Tiny refactor and make a visible check for count so that it is not flaky in local

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Fix import paths after merging the develop

yash-learner committed:
Add test to verify content in activity definition table row

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Update based on new changes in the UI by recent PR merge

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Revert changes made to the pw config

nihal467 committed:
Merge branch 'develop' into add-tests-for-labs

nihal467 commented:
@yash-learner the playwright runner is failing, seem like a flaky test exist in the PR, can you investigate the failure and fix it

yash-learner committed:
Update text based on change in PR#14308

-  Add video retention on first retry

yash-learner committed:
Revert change to retries

yash-learner committed:
Update text for other test as well

yash-learner commented on tests/facility/settings/activityDefinitions/activityDefinitionList.spec.ts:
```diff
@@ -0,0 +1,185 @@
+    await expect(
+      page.getByText(/showing 1 of 1 activity definitions/i),
+    ).toBeVisible();
+  });
```
I went to specific test because it is search and it based on backend seeds 

At the time, I was writing this code, I didn't feel to add more code for creating a AD just for search and then search it!

nihal467 commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,529 @@
+    await page.getByLabel(/usage.*\*/i).fill(testData.usage);
+
+    await page.getByLabel(/^status$/i).click();
+    await page.getByRole("option", { name: /active/i }).click();
```
- use the faker array with all the option in the status dropdown and randomly select it

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/14108/commits/2535e7249f33ce1ac323f874be1dfeba8f07b56a

nihal467 commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,529 @@
+    await page.getByRole("option", { name: /active/i }).click();
+
+    await page.getByRole("combobox", { name: /^category\s*\*$/i }).click();
+    await page.getByRole("option").first().click();
```
- use the faker array and random selection

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/14108/commits/2535e7249f33ce1ac323f874be1dfeba8f07b56a

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,529 @@
+    await page.getByLabel(/^kind$/i).click();
+    await page.getByRole("option", { name: /service request/i }).click();
+
+    const codeCombobox = page.getByRole("combobox", { name: /^code/i });
```
create a faker array with 10 random option from the dropdown and select the activity code based on that

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/14108/commits/2535e7249f33ce1ac323f874be1dfeba8f07b56a

Using rand int for this to pick from options -> this as well randomizes

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,529 @@
+    await expect(technicalDetailsCard).toBeVisible();
+    await expect(
+      technicalDetailsCard.getByText("Service Request"),
+    ).toBeVisible();
```
- add visibility check for the randomly selected array

yash-learner replied:
<img width="720" height="274" alt="image" src="https://github.com/user-attachments/assets/9bc6dcd3-fd87-4298-b7f6-b6be85d3c668" />

Right now we have only on option for kind `Service Request`

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,529 @@
+        ).toBeVisible();
+
+        await page.getByLabel(/title.*\*/i).fill(testData.title);
+        await page.getByLabel(/description.*\*/i).fill(testData.description);
```
- verify the slug auto filling based on the title

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/14108/commits/8b2c9a1b0d4ab4e975142c42eded5c0dd520389c

nihal467 commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,529 @@
+        await page.getByLabel(/usage.*\*/i).fill(testData.usage);
+
+        await page.getByLabel(/^status$/i).click();
+        await page.getByRole("option", { name: /active/i }).click();
```
- use faker array

yash-learner replied:
https://github.com/ohcnetwork/care_fe/commit/2535e7249f33ce1ac323f874be1dfeba8f07b56a

nihal467 commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,529 @@
+        await page.getByRole("option", { name: /active/i }).click();
+
+        await page.getByRole("combobox", { name: /^category\s*\*$/i }).click();
+        await page.getByRole("option").first().click();
```
- use faker array

yash-learner replied:
https://github.com/ohcnetwork/care_fe/commit/2535e7249f33ce1ac323f874be1dfeba8f07b56a

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,529 @@
+        await page.getByRole("option", { name: /service request/i }).click();
+
+        const codeCombobox = page.getByRole("combobox", { name: /^code/i });
+        await selectFromValueSet(page, codeCombobox, {
```
use faker array

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/14108/commits/2535e7249f33ce1ac323f874be1dfeba8f07b56a

Using rand int for this to pick from options -> this as well randomizes

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,529 @@
+
+      await test.step("select diagnostic report codes", async () => {
+        const diagSection = page
+          .getByText(/^diagnostic report codes$/i)
```
- use faker array here

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/14108/commits/2535e7249f33ce1ac323f874be1dfeba8f07b56a

Using rand int for this to pick from options -> this as well randomizes

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,529 @@
+        `/facility/${facilityId}/settings/activity_definitions`,
+      );
+
+      await test.step("navigate and verify details", async () => {
```
- verify the dynamically selected values

nihal467 commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,529 @@
+
+    const updatedData = generateActivityDefinitionData();
+
+    await test.step("edit activity definition", async () => {
```
- Move editActivitDefinition to a separate file, and lets keep it independent from already created activity

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/14108/commits/2535e7249f33ce1ac323f874be1dfeba8f07b56a

nihal467 commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,529 @@
+      );
+    });
+
+    await test.step("verify updated details in view page", async () => {
```
- keep it as a separate test

yash-learner replied:
It was added based on Bodhi's comment https://github.com/ohcnetwork/care_fe/pull/14108#discussion_r2462266765

nihal467 commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,529 @@
+        ).toBeVisible();
+      });
+
+      await test.step("verify prefilled fields", async () => {
```
keep it as a separate test

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/14108/commits/2535e7249f33ce1ac323f874be1dfeba8f07b56a

nihal467 commented on tests/facility/settings/activityDefinitions/activityDefinitionForm.spec.ts:
```diff
@@ -0,0 +1,529 @@
+      ).toBeVisible();
+    });
+
+    await test.step("delete activity definition", async () => {
```
- keep it as a separate test in new test file actiityDefinitionDeletion

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/14108/commits/2535e7249f33ce1ac323f874be1dfeba8f07b56a

nihal467 commented on tests/facility/settings/activityDefinitions/activityDefinitionList.spec.ts:
```diff
@@ -0,0 +1,185 @@
+    ).toBeVisible();
+  });
+
+  test("should filter by status and clear filter", async ({ page }) => {
```
- verify that all the status are showing some files based on there status

nihal467 commented on tests/facility/settings/activityDefinitions/activityDefinitionList.spec.ts:
```diff
@@ -0,0 +1,185 @@
+    ).toBeVisible();
+  });
+
+  test("should filter by status and clear filter", async ({ page }) => {
```
- verify the status on each activity definition by checking some definition are coming

yash-learner replied:
Moving list specs into a separate PR

nihal467 commented on tests/facility/settings/activityDefinitions/activityDefinitionList.spec.ts:
```diff
@@ -0,0 +1,185 @@
+    ).toBeVisible();
+  });
+
+  test("should filter by category/classification and change filters", async ({
```
- same here, verify some data are present based on the category classification inthe list for all positon in the dropdown of category

yash-learner replied:
Moving list specs into a separate PR

nihal467 commented on tests/facility/settings/activityDefinitions/activityDefinitionList.spec.ts:
```diff
@@ -0,0 +1,185 @@
+      ).toBeVisible();
+    });
+
+    await test.step("verify classification", async () => {
```
- why are making all single visibility check a step, ?

yash-learner replied:
Moving list specs into a separate PR, Will fix in the separate PR

nihal467 commented on tests/facility/settings/activityDefinitions/categoryList.spec.ts:
```diff
@@ -0,0 +1,148 @@
+      const searchInput = page.getByPlaceholder(/search categories/i);
+      await expect(searchInput).toBeVisible();
+
+      await searchInput.fill(testData.name);
```
- dont make dependent test, which can lead for flakiness

yash-learner replied:
Moving list specs into a separate PR, Will fix in the separate PR

nihal467 commented on tests/facility/settings/activityDefinitions/categoryList.spec.ts:
```diff
@@ -0,0 +1,148 @@
+      await expect(addActivityButton).toBeEnabled();
+    });
+
+    test("should edit category", async ({ page }) => {
```
- add a one more test to verify the error in the form

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Randomise the status and classification

Renama categoryName to resourceCategotyName

yash-learner committed:
Address review comments

- Separate edit and delete files
- Separate test cases for verify details
- Helper file to generate relevant data for specs
- Randomise
- More

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Merge branch 'add-tests-for-labs' of https://github.com/ohcnetwork/care_fe into add-tests-for-labs

yash-learner committed:
Check slug auto generation and some refactor in edit

yash-learner committed:
Quick test to test timeout

- Add some AD to check data for all status and classification

yash-learner committed:
Refactor activity definition list tests to use parameterized status and classification filtering

yash-learner committed:
Moving list specs to a separate follow up PR

yash-learner commented on tests/facility/settings/activityDefinition/activityDefinitionEdit.spec.ts:
```diff
@@ -0,0 +1,412 @@
+});
+
+test.describe("activity definition edit", () => {
+  test("should display all prefilled fields", async ({ page }) => {
```
This is only the case where dynamic names are not used because for all the other cases i am using `faker.number.int()` to select dynamically from options without faker array but I here I am do exact text of the specimen, observation prefilled so need to store text and did feel worth to create faker array for this one case.

yash-learner replied:
updated to faker array https://github.com/ohcnetwork/care_fe/pull/14108/commits/7da61dbeae1b4580276e0c146d04f2def91bf9b0

yash-learner committed:
Add case for validation errors

yash-learner commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,362 @@
+    const codeCombobox = page.getByRole("combobox", { name: /^code/i });
+    await selectFromValueSet(page, codeCombobox, {
+      itemIndex: faker.number.int({ min: 0, max: 4 }),
+    });
```
Those min and max values choose based on BE seeds available

yash-learner replied:
For valuesets it does not matter because the list way larger

yash-learner committed:
Remove duplicate visibility check for slug required error in activity definition creation tests

yash-learner commented on tests/facility/settings/activityDefinitions/activityDefinitionEdit.spec.ts:
```diff
@@ -0,0 +1,412 @@
+      observationRequirement: "glucose",
+      chargeItem: "test",
+      location: "Pharmacy",
+      diagnosticReport: "lab",
```
https://github.com/ohcnetwork/care_fe/pull/14108#discussion_r2545015122

yash-learner commented on tests/facility/settings/activityDefinitions/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,360 @@
+    await expect(
+      technicalDetailsCard.getByText("Service Request"),
+    ).toBeVisible();
+  });
```
In detail check will be done in list spec,

Those headings will show up when there is at least one item.

yash-learner replied:
Also pre-filled fields test verifies this case right ?

Whether already saved data is present or not ?

yash-learner committed:
Move to faker arrayElement from faker number int for SD, OD, Location, CID, DR

- Update  the tests

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Add copilot suggestions

yash-learner committed:
Merge branch 'add-tests-for-labs' of https://github.com/ohcnetwork/care_fe into add-tests-for-labs

yash-learner committed:
Handle multi text match

yash-learner commented on tests/facility/settings/activityDefinitions/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,399 @@
+    await clearButton.click();
+
+    const searchInput = page.getByPlaceholder(/search/i);
+    await searchInput.fill(testData.title);
```
https://github.com/ohcnetwork/care_fe/pull/14108/commits/56ce24b129607bef90506703cc69fced0dbc7487

yash-learner commented on tests/facility/settings/activityDefinitions/activityDefinitionDelete.spec.ts:
```diff
@@ -0,0 +1,76 @@
+    await clearButton.click();
+
+    const searchInput = page.getByPlaceholder(/search/i);
+    await searchInput.fill(createdAD.title);
```
https://github.com/ohcnetwork/care_fe/pull/14108/commits/56ce24b129607bef90506703cc69fced0dbc7487

yash-learner commented on tests/facility/settings/activityDefinitions/activityDefinitionEdit.spec.ts:
```diff
@@ -0,0 +1,429 @@
+      .getByRole("button")
+      .filter({ has: page.locator("svg.lucide-x") })
+      .first();
+    await clearButton.click();
```
https://github.com/ohcnetwork/care_fe/pull/14108/commits/56ce24b129607bef90506703cc69fced0dbc7487

yash-learner commented on tests/helpers/activityDefinition.ts:
```diff
@@ -0,0 +1,195 @@
+  const testData = {
+    title:
+      options.overrides?.title ||
+      `${faker.science.chemicalElement().name.slice(0, 16)}_${faker.string.uuid().slice(0, 8)}`,
```
https://github.com/ohcnetwork/care_fe/pull/14108/commits/56ce24b129607bef90506703cc69fced0dbc7487

yash-learner commented on tests/helpers/activityDefinition.ts:
```diff
@@ -0,0 +1,195 @@
+    Object.values(Classification),
+  );
+  return {
+    title: `${faker.science.chemicalElement().name.slice(0, 16)}_${faker.string.uuid().slice(0, 8)}`,
```
https://github.com/ohcnetwork/care_fe/pull/14108/commits/56ce24b129607bef90506703cc69fced0dbc7487

yash-learner commented on playwright.config.ts:
```diff
@@ -13,11 +13,13 @@ import { defineConfig, devices } from "@playwright/test";
  */
 export default defineConfig({
   testDir: "./tests",
+
```
Not required

yash-learner committed:
Push missed out changes

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner commented on tests/facility/settings/activityDefinition/activityDefinitionEdit.spec.ts:
```diff
@@ -0,0 +1,422 @@
+  test("should display all prefilled fields", async ({ page }) => {
+    await page.goto(`/facility/${facilityId}/settings/activity_definitions`);
+    await page.getByText(resourceCategoryName).click();
+
```
https://github.com/ohcnetwork/care_fe/pull/14108/commits/41a439e7a4aee965cd5373f78d531d24718520c5

yash-learner committed:
Rename folder name

yash-learner committed:
Merge branch 'add-tests-for-labs' of https://github.com/ohcnetwork/care_fe into add-tests-for-labs

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Use clearFilter

yash-learner committed:
Merge branch 'add-tests-for-labs' of https://github.com/ohcnetwork/care_fe into add-tests-for-labs

yash-learner commented on tests/facility/settings/activityDefinition/activityDefinitionDelete.spec.ts:
```diff
@@ -0,0 +1,72 @@
+      .getByRole("button")
+      .filter({ has: page.locator("svg.lucide-x") })
+      .first()
+      .click();
```
https://github.com/ohcnetwork/care_fe/pull/14108/commits/b2cfaa75c483cb3d197319d0b690392e023f6340

nihal467 commented on tests/helpers/activityDefinition.ts:
```diff
@@ -0,0 +1,185 @@
+import {
+  Classification,
+  Status,
+} from "src/types/emr/activityDefinition/activityDefinition";
```
```suggestion
```

- lets not do this, lets have our own static list, so that incase the code breaks, the test also should fail

yash-learner replied:
code is using the same enum right ?

https://github.com/ohcnetwork/care_fe/blob/develop/src/pages/Facility/settings/activityDefinition/ActivityDefinitionForm.tsx#L52-L54

so if code fails, this also fail right ?

🤔

nihal467 replied:
If the test uses the same enum as the code, it cannot catch unintended enum changes. Tests must be independent

yash-learner replied:
if enum changes then tests should fail right, that's how we get to know we need to update the tests ?

nihal467 commented on tests/helpers/activityDefinition.ts:
```diff
@@ -0,0 +1,185 @@
+    Object.values(Classification),
+  );
+  return {
+    title: `${faker.science.chemicalElement().name.slice(0, 16)}_${faker.string.uuid().slice(0, 8)}`,
```
dont make this title, complicated

yash-learner replied:
simplified https://github.com/ohcnetwork/care_fe/pull/14108/commits/24d4c5ae4f709af638076d77286ca979cce2b048#diff-2ccdcce2eb8f890e1732b8332bb2454f5329c9b08821e622ae7b9619de1f2cf9

nihal467 commented on tests/helpers/activityDefinition.ts:
```diff
@@ -0,0 +1,185 @@
+ * @param title - The title to convert to a slug
+ * @returns The expected slug value
+ */
+export function generateExpectedSlug(title: string): string {
```
can't you use the faker uuid and manipulate it, rather than keeping a custom generator

yash-learner replied:
I am using faker to generate tittle and replace spaces with `-` which already done in other tests like chargeItem

I extracted into a function so that I can reuse across tests

nihal467 commented on tests/helpers/activityDefinition.ts:
```diff
@@ -0,0 +1,185 @@
+  const resourceCategoryName = options.resourceCategoryName || "Lab Tests";
+
+  // Generate test data with random values, then apply any overrides
+  const testData = {
```
- why are you creating a generator, then fill a form, in using the data to overwrite existing variable, cant we keep it simple and directly use faker and assign value to variable

nihal467 commented on tests/helpers/activityDefinition.ts:
```diff
@@ -0,0 +1,185 @@
+
+  await page.getByLabel(/^status$/i).click();
+  await page
+    .getByRole("option", { name: new RegExp(testData.status, "i") })
```
- when you remove the code import , this should be simplified

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinition.ts:
```diff
@@ -0,0 +1,185 @@
+  await page.getByRole("option", { name: /service request/i }).click();
+
+  const codeCombobox = page.getByRole("combobox", { name: /^code/i });
+  await selectFromValueSet(page, codeCombobox, {
```
```suggestion
```

- cant we just use faker with custom list and randomly select one from the list after searching

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,394 @@
+  SPECIMEN_DEFINITIONS,
+  generateActivityDefinitionData,
+  generateExpectedSlug,
+} from "tests/helpers/activityDefinition";
```
```suggestion
```

- lets not keep the form creation outside the activity defintion folder

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,394 @@
+      .click();
+    await page.getByRole("button", { name: /^create$/i }).click();
+
+    await expect(page.getByText(/title.*required/i)).toBeVisible();
```
```suggestion
```

verify the form label id and verify the error message is for that label

nihal467 replied:
this need to done for all the visibility check

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,394 @@
+  test("should create activity definition with required fields", async ({
+    page,
+  }) => {
+    const testData = generateActivityDefinitionData();
```
- you have a entire helper function to create definition, then why are you duplicating this

yash-learner replied:
https://github.com/ohcnetwork/care_fe/pull/14108#discussion_r2559706334

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionCreate.spec.ts:
```diff
@@ -0,0 +1,394 @@
+      DIAGNOSTIC_REPORT_CODES,
+    );
+
+    await test.step("create activity definition", async () => {
```
why are you not reusing the helper function

yash-learner replied:
The helper function will just create the AD with basic fields 

Here  all fields are filled and that helper mostly used in edit, delete, list page to test different things

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionDelete.spec.ts:
```diff
@@ -0,0 +1,68 @@
+    await page.goto(`/facility/${facilityId}/settings/activity_definitions`);
+    await page.getByText(resourceCategoryName).click();
+
+    await clearFilter(page);
```
- why are you callig clear filter for the delete activity definition

yash-learner replied:
To verify the deleted AD is visible with only retried status badge 

by default some filters applied

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionEdit.spec.ts:
```diff
@@ -0,0 +1,422 @@
+
+    const obsContainer = page
+      .getByText(/^observation requirements$/i)
+      .locator("..");
```
```suggestion

use proper locator

nihal467 commented on tests/facility/settings/activityDefinition/activityDefinitionEdit.spec.ts:
```diff
@@ -0,0 +1,422 @@
+      specimenContainer2.getByText(selectedSpecimen).first(),
+    ).toBeVisible();
+
+    const obsContainer2 = page
```
- why are you using a const for a single action, cant you directly use it and you proper location

yash-learner committed:
Refactor slug generation: move generateExpectedSlug to utils and update activity definition tests to use the new import. Clean up LOCATIONS array in activityDefinition helper.

yash-learner committed:
wip: Address review comments

Refactor activity definition tests: streamline creation logic by removing resourceCategoryName parameter,
update slug generation to use expectedSlug utility, and
enhance test data generation for improved clarity and maintainability.

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Update validation error case to utilize getFieldErrorMessage

yash-learner committed:
Move ad helper from helper folder to under activityDefinition folder

yash-learner committed:
Refactor activity definition data generation and update related tests

- Updated `generateActivityDefinitionData` to include optional fields and improved structure.
- Refactored `createActivityDefinition` to utilize the new data generation function.
- Adjusted Playwright tests for activity definition creation, editing, and deletion to align with the updated data structure and navigation paths.
- Removed unused `clearFilter` helper function from UI tests.

yash-learner committed:
Fix path

yash-learner committed:
Refactor Playwright tests to use getCardByTitle helper for card element selection

- Replaced direct locator usage with the new `getCardByTitle` function in both activity definition create and edit tests for improved readability and maintainability.
- Added `getCardByTitle` function in the UI helper to streamline card selection by title.

yash-learner committed:
Fix the js doc

yash-learner committed:
Assert slug instead of fill in CAD helper

yash-learner committed:
Tiny change: Update test descriptions a bit

yash-learner committed:
Address a coderabbit nitpick

- Simplified the return statement in `createActivityDefinition` by directly returning the `data` object, removing unnecessary destructuring of fields.

yash-learner committed:
Address coderabbit suggestion

yash-learner committed:
Fill slug in edit

yash-learner commented on tests/facility/settings/activityDefinition/activityDefinitionEdit.spec.ts:
```diff
@@ -0,0 +1,258 @@
+        .filter({
+          has: page.locator("label", {
+            hasText: /^observation requirements$/i,
+          }),
+        })
+        .getByText(createdAD.observation!)
+        .first(),
+    ).toBeVisible();
+
+    await expect(
+      page
+        .locator("div.rounded-lg")
+        .filter({
+          has: page.locator("label", {
+            hasText: /^charge item definitions$/i,
+          }),
+        })
+        .getByText(createdAD.chargeItem!)
+        .first(),
+    ).toBeVisible();
+
+    await expect(
+      page
+        .locator("div.rounded-lg")
+        .filter({ has: page.locator("label", { hasText: /^locations$/i }) })
+        .getByText(createdAD.location!)
+        .first(),
+    ).toBeVisible();
+
+    await expect(
+      page
+        .locator("div.rounded-lg")
+        .filter({
+          has: page.locator("label", {
+            hasText: /^diagnostic report codes$/i,
+          }),
+        })
+        .getByText(createdAD.diagnosticReportCode!)
+        .first(),
+    ).toBeVisible();
```
I know but that's the only way here

yash-learner committed:
Move ui.ts, utils.ts from helpers to helper
